### PR TITLE
[LWD] feat(baanx-client): focus scope on login with minimal coverage

### PR DESCRIPTION
**Pull request by martijnhjk**
### 📝 Description

A private, unpublished helper offering headless Baanx login through the CLI and programatically in test.

### 🔗 Context

- **JIRA / GitHub issue**: https://ledgerhq.atlassian.net/browse/QAA-1508
- **ADR** (if any):

### CODE REVIEW

File: .changeset/baanx-test-client-library.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+"@ledgerhq/baanx-test-client": minor
+---
+
+Add `@ledgerhq/baanx-test-client`: a test-only client that logs a sandbox test user in against `dev.api.baanx.com` (injectable base URL). Branches on the response body rather than the status code (Baanx answers 200 even when login has not completed), answers the OTP challenge by deriving the code from the user's TOTP setup key, and returns a cached access token plus the metadata needed to debug a failure.
+
+Exposes `getBaanxAuthToken()` for Playwright/Detox/Jest setup and a CLI (`pnpm --silent --filter @ledgerhq/baanx-test-client token`) for Postman, curl and CI. Tokens are cached in-process and shared between concurrent callers so a parallel run does not log in per worker; credentials come from `BAANX_TEST_*` environment variables (`BAANX_TEST_CLIENT_KEY`, `BAANX_TEST_API_URL`, user email/password/TOTP), not the app's `CARD_BAANX_*`, and nothing is logged or written to disk.
+
+Includes an opt-in live integration test (`test-integ`) that logs in against the sandbox; it is excluded from the hermetic unit run and skips when credentials are not configured.
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply .changeset/baanx-test-client-library.md]: pushed!

File: e2e/tooling/baanx-test-client/README.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+`{ accessToken, expiresIn, refreshToken }` — through the OAuth/PKCE flow in
+`@features/flow-pay-card-auth`, and stores it in `@features/platform-card`. What this package does is
+a different flow entirely: password + TOTP, straight against Baanx. The token is a `Bearer`
+credential for the Baanx API; it is not an equivalent of `cardSession` and cannot become one, because
+the password login returns no refresh token.
+
+There is one deliberate bridge, and it is a test scaffold rather than an equivalence:
+[`--session`](#starting-ledger-wallet-desktop-already-signed-in) synthesises a `PayCardSession` with
+a placeholder refresh token so Desktop can be started already signed in for local development and
+E2E. Read that section before using it, in particular the one about `CARD_API_URL` having to match
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/README.md]: pushed 👍

File: pnpm-lock.yaml [hunk 3]
```diff
@@ -43786,7 +43830,7 @@ snapshots:
 
   '@babel/template@7.27.2':
     dependencies:
-      '@babel/code-frame': 7.29.7
+      '@babel/code-frame': 7.29.0
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply pnpm-lock.yaml]: pushed 🐳

File: shared/env/src/definitions/team-qaa/index.ts [hunk 4]
```diff
@@ -11,6 +11,13 @@ const teamQaa = {
     parser: boolParser,
     desc: "true when launched for E2E testing",
   },
+  CARD_SESSION_BOOTSTRAP: {
+    def: "",
+    parser: stringParser,
+    desc:
+      "Test-only. A PayCardSession as JSON, injected at launch so the app starts signed in. " +
+      "Read in a dev build, or in any build launched with PLAYWRIGHT_RUN — a packaged one included.",
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply shared/env/src/definitions/team-qaa/index.ts]: pushed !

File: e2e/tooling/baanx-test-client/README.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+### Starting Ledger Wallet Desktop already signed in
+
+Desktop cannot finish the OAuth login today — the hosted page opens in the user's own browser and
+reports nothing back (LIVE-34740) — so a signed-in state has to be injected. `-- --session` prints a
+`PayCardSession` for exactly that:
+
+```bash
+export CARD_SESSION_BOOTSTRAP=$(pnpm --silent --filter @ledgerhq/baanx-test-client token -- --session)
+export CARD_API_URL=https://<the Baanx host that minted the token>
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/README.md]: already pushed

File: pnpm-lock.yaml [hunk 6]
```diff
@@ -43819,7 +43863,7 @@ snapshots:
 
   '@babel/traverse@7.28.5':
     dependencies:
-      '@babel/code-frame': 7.29.7
+      '@babel/code-frame': 7.29.0
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply pnpm-lock.yaml:43879]: pushed

File: e2e/tooling/baanx-test-client/src/auth/login.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  if (asRecord(retry.body).isOtpRequired === true) {
+    throw new BaanxOtpError(
+      "Baanx still requires an OTP after the generated code was submitted. The code was rejected — " +
+        `check that ${ENV_VARS.totpSecret} is this user's setup key and that the digits/period/algorithm ` +
+        `match how the authenticator was enrolled.`,
+      redactBody(retry.body, [...secretsOf(config), code]),
+    );
+  }
+
+  return toSession(config, retry.body, { otpUsed: true });
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/login.ts]: pushed!

File: .changeset/pay-card-session-bootstrap.md [hunk 8]
```diff
@@ -0,0 +1,10 @@
+---
+"ledger-live-desktop": patch
+"@shared/env": patch
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply .changeset/pay-card-session-bootstrap.md]: pushed

File: e2e/tooling/baanx-test-client/src/http/body.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+ * Replace every occurrence of a known secret in a string.
+ *
+ * No minimum length: a short credential is still a credential, and a mangled
+ * message is a better outcome than a leaked one. Lives here rather than in
+ * `send.ts` so body and message redaction share exactly one implementation.
+ */
+export function redactSecretsInText(text: string, secrets: readonly string[]): string {
+  let out = text;
+  for (const secret of secrets) {
+    if (secret) out = out.split(secret).join(REDACTION_PLACEHOLDER);
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/http/body.ts]: pushed!

File: apps/ledger-live-desktop/src/renderer/bootstrapCardSession.test.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      "accessToken must be a string",
+    );
+  });
+});
+
+describe("bootstrapCardSession", () => {
+  const dispatch = jest.fn();
+
+  beforeEach(() => {
+    delete process.env.CARD_SESSION_BOOTSTRAP;
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply apps/ledger-live-desktop/src/renderer/bootstrapCardSession.test.ts:58]: pushed

File: e2e/tooling/baanx-test-client/README.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+### Starting Ledger Wallet Desktop already signed in
+
+Desktop cannot finish the OAuth login today — the hosted page opens in the user's own browser and
+reports nothing back (LIVE-34740) — so a signed-in state has to be injected. `-- --session` prints a
+`PayCardSession` for exactly that:
+
+```bash
+export CARD_SESSION_BOOTSTRAP=$(pnpm --silent --filter @ledgerhq/baanx-test-client token -- --session)
+export CARD_BAANX_API_URL=https://<the Baanx host that minted the token>
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/README.md:105]: pushed

File: apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+ *
+ * The gate is a runtime check rather than a build-time constant (`__DEV__` / a `TESTING` define) on
+ * purpose. Release-mode E2E runs against the release bundle — the workflow's `build_type: js`, which
+ * its own input describes as "pick js for release testing" — and a build-time marker is by definition
+ * absent there, so a compile-time gate would silently disable session injection in exactly the run
+ * that most needs it. Please don't "harden" this to a define without solving that first.
+ */
+export async function bootstrapCardSession(dispatch: AppDispatch): Promise<void> {
+  const raw = process.env.CARD_SESSION_BOOTSTRAP;
+  if (raw) delete process.env.CARD_SESSION_BOOTSTRAP;
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts:34]: The bootstrap only executes if app runs in test mode explicitly.

We want to keep it in the parent process deliberately so that relaunches re-inject the value.

Ignoring this comment.

File: e2e/tooling/baanx-test-client/package.json [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    "otpauth": "catalog:"
+  },
+  "devDependencies": {
+    "@jest/globals": "catalog:",
+    "@swc/core": "catalog:",
+    "@swc/jest": "catalog:",
+    "@types/jest": "catalog:",
+    "@types/node": "catalog:",
+    "jest": "catalog:",
+    "tsx": "catalog:",
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/package.json:44]: pushed

File: e2e/tooling/baanx-test-client/src/auth/totp.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+  let at = startedAt;
+  let window = totpWindow(config.period, at);
+
+  if (window.remainingMs < minRemainingMs) {
+    // Sleeping exactly the remainder lands us in the next window with a full
+    // period ahead. Re-read the clock rather than assuming the sleep was exact.
+    await clock.sleep(window.remainingMs);
+    at = clock.now();
+    window = totpWindow(config.period, at);
+  }
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/totp.ts:86]: Clamping to config.period * 1_000 makes the threshold equal to a full window. remainingMs only hits that on the exact boundary; one millisecond later it is already smaller, so while (remainingMs < minRemaining) can keep sleeping. A loop that re-reads the clock is the right idea, but the floor needs to be strictly less than the period (or we reject periods shorter than 2s) - ignoring this particular suggestion.

File: e2e/tooling/baanx-test-client/src/http/send.ts [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+  let response: Response;
+  try {
+    response = await fetchImpl(`${baseUrl}${path}`, {
+      method: "POST",
+      headers,
+      body: JSON.stringify(body),
+    });
+  } catch (error) {
+    throw new BaanxTransportError(baseUrl, transportReason(error));
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/http/send.ts]: pushed.

File: e2e/tooling/baanx-test-client/src/http/send.ts [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+  switch (response.status) {
+    case 498:
+      return new BaanxInvalidClientKeyError(apiMessage, ENV_VARS.clientKey);
+    case 499:
+      return new BaanxMissingClientKeyError(apiMessage, ENV_VARS.clientKey);
+    case 401:
+      return new BaanxInvalidCredentialsError(apiMessage, looksAccountLocked(apiMessage));
+    case 429:
+      return new BaanxRateLimitError(apiMessage, response.retryAfter);
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/http/send.ts]: Retry-After is not a secret channel. We only send the client key, password, and OTP in the request header/body, and we already scrub those when they come back in a body — which this API actually does. RFC Retry-After is a delay or a date; nothing we call, and nothing Baanx documents, puts credentials there. Redacting it would only matter if the header literally contained those strings, which a normal 30 or HTTP-date will not.

File: e2e/tooling/baanx-test-client/src/auth/session.ts [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    .then(session => {
+      cache.set(key, { session });
+      return session;
+    })
+    .catch((error: unknown) => {
+      // Drop the failed attempt so the next caller retries rather than
+      // awaiting a promise that is already rejected.
+      cache.delete(key);
+      throw error;
+    });
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/session.ts]: pushed

File: e2e/desktop/tests/specs/paytab.spec.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
+import { test } from "tests/fixtures/common";
+import { Team } from "@ledgerhq/live-e2e-shared/enum/Team";
+import { FF_LWD_CONTACTS, FF_LWD_PAY_TAB } from "tests/utils/featureFlagUtils";
+import { DEVICE_TAGS } from "tests/utils/tagsUtils";
+
+// Skipped: proves the Baanx client and token injection work; not a Pay tab UI assertion yet.
+test.describe.skip("Pay tab", () => {
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/desktop/tests/specs/paytab.spec.ts:7]: this is intentional, to be unskipped soon...

File: e2e/tooling/baanx-test-client/src/auth/login.ts [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+  deps: LoginDeps = {},
+): Promise<BaanxAuthSession> {
+  const fetchImpl = deps.fetchImpl ?? globalThis.fetch;
+  const { email, password } = config;
+
+  const first = await post(config, LOGIN_PATH, { email, password }, fetchImpl);
+
+  const phase = asString(asRecord(first.body).phase);
+  if (phase) {
+    throw new BaanxOnboardingIncompleteError(phase, asString(asRecord(first.body).userId));
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/login.ts:46]: pushed

File: e2e/desktop/tests/specs/paytab.spec.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/desktop/tests/specs/paytab.spec.ts:7]: intentional, ignoring!

File: e2e/tooling/baanx-test-client/src/cliArgs.test.ts [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+    expect(parseCliArgs([])).toEqual({ kind: "run", format: "token" });
+  });
+
+  it("recognises --json", () => {
+    expect(parseCliArgs(["--json"])).toEqual({ kind: "run", format: "json" });
+  });
+
+  it("recognises --session", () => {
+    expect(parseCliArgs(["--session"])).toEqual({ kind: "run", format: "session" });
+  });
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/cliArgs.test.ts:14]: intentional - contributors requested less coverage in previous rounds.

File: e2e/tooling/baanx-test-client/package.json [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+  "license": "Apache-2.0",
+  "scripts": {
+    "token": "tsx src/cli.ts",
+    "typecheck": "tsc --noEmit",
+    "test": "jest",
+    "test:watch": "jest --watch",
+    "coverage": "jest --coverage",
+    "lint": "oxlint ./src",
+    "lint:fix": "pnpm lint --fix",
+    "unimported": "pnpm knip --directory ../../.. -W e2e/tooling/baanx-test-client"
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/package.json:32]: pushed

File: e2e/tooling/baanx-test-client/project.json [hunk 22]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "@ledgerhq/baanx-test-client",
+  "targets": {
+    "build": {
+      "executor": "nx:noop"
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/project.json]: Nx infers test from the package’s "test": "jest" script (nx:run-script), which is the same pattern as other libs (e.g. @shared/auth). Confirmed with nx show project @ledgerhq/baanx-test-client: test is present, tagged scope:e2e-tooling, and build is nx:noop so the default test → build dependency is fine.

File: e2e/tooling/baanx-test-client/src/config.ts [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+  if (missing.length > 0) throw new BaanxConfigError(missing);
+
+  const baseUrl =
+    clean(overrides.baseUrl, ENV_VARS.baseUrl) ??
+    read(env, ENV_VARS.baseUrl) ??
+    DEFAULT_BAANX_BASE_URL;
+
+  return {
+    // A trailing slash would double up against the "/v1/..." paths.
+    baseUrl: trimTrailing(baseUrl, "/"),
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/config.ts:69]: debunking the priority on this - we are only passing test credentials

File: e2e/tooling/baanx-test-client/src/auth/login.ts [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+import {
+  BaanxHttpError,
+  BaanxNoTokenError,
+  BaanxOnboardingIncompleteError,
+  BaanxOtpError,
+} from "../errors";
+import { resolveExpiry } from "./expiry";
+import { sendJson, toTypedError } from "../http/send";
+import type { BaanxResponse } from "../http/send";
+import { generateFreshTotpCode, systemClock } from "./totp";
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/login.ts]: This looks like a stale or swapped reading of the two modules. On this branch login.ts already imports resolveExpiry from ./expiry and the TOTP helpers from ./totp, which is where those exports live. pnpm --filter @ledgerhq/baanx-test-client typecheck passes. Happy to point at the exact lines if you’re looking at an older commit.

File: e2e/tooling/baanx-test-client/package.json [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "exports": {
+    ".": "./src/index.ts",
+    "./package.json": "./package.json"
+  },
+  "sideEffects": false,
+  "license": "Apache-2.0",
+  "scripts": {
+    "token": "tsx src/cli.ts",
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/package.json:23]: Knip already treats src/cli.ts as an entry: "token": "tsx src/cli.ts" is picked up by the tsx plugin (deferResolveEntry:src/cli.ts). pnpm --filter @ledgerhq/baanx-test-client unimported is green, so this does not fail codecheck. A bin field would still be nicer packaging, but it is not required for Knip.

File: e2e/tooling/baanx-test-client/src/auth/login.ts [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+        `Underlying failure: ${error.message}`,
+      error.body,
+    );
+  }
+}
+
+/** Values that must never survive into an error message. */
+function secretsOf(config: ResolvedBaanxAuthConfig): readonly string[] {
+  return [config.clientKey, config.password, config.totp.secret];
+}
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/login.ts:152]: fix pushed

File: apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+
+  if (!__DEV__ && !getEnv("PLAYWRIGHT_RUN")) return;
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts]: pushed

File: e2e/desktop/tests/specs/paytab.spec.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/desktop/tests/specs/paytab.spec.ts:7]: intentional, ignore

File: apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+  if (!raw) return;
+
+  try {
+    const session = parseSession(raw);
+    await cardSession.set(session);
+    dispatch(setSignedIn(true));
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts:44]: This is an edge case that can be improved in the future. For now manual testers can log out in the app and e2e will use a clear app session every time.

File: e2e/tooling/baanx-test-client/src/cli.ts [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+  const expiresIn = Math.max(
+    1,
+    Math.floor((Date.parse(session.expiresAt) - Date.parse(session.issuedAt)) / 1000),
+  );
+
+  // Matches the app's PayCardSession exactly: accessToken, refreshToken, expiresIn.
+  return JSON.stringify({
+    accessToken: session.accessToken,
+    refreshToken: REFRESH_TOKEN_PLACEHOLDER,
+    expiresIn,
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/cli.ts:95]: US case not required for e2e right now - can be implemented as a future improvement, ignore for now.

File: pnpm-lock.yaml [hunk 30]
```diff
@@ -23194,7 +23241,6 @@ packages:
     resolution: {integrity: sha512-KRbkSthClEwkbpt/LLJmBJhIOvOsyrp9OICisF9p3mOODjaxAuYmyOgrVreg09BT2F4hXN3JXpvt9eIZpTCRsw==}
     cpu: [x64]
     os: [linux]
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply pnpm-lock.yaml:23263]: pushed a change

**Comment by martijnhjk** [comment]
Copilots minor comments should not block this PR - please find a response to each item below:

bootstrapCardSession.ts:35 — not critical
if (raw) only skips delete when the value is already empty/undefined. An explicit CARD_SESSION_BOOTSTRAP="" is not a bearer; getAllEnvs() still never sees it. Clearing exists to drop a real payload after read. Unconditional delete would match the comment more literally, but it is not a security hole.

README.md:148 — false positive
Both spans are closed on their own line (`null`, then `Bearer null`). CommonMark does not leave a span open across that wrap; the next line is a normal list continuation.

README.md:110 — not critical
The block under that sentence already exports all three. “Both” is “Playwright needs the same env as start”, not a two-variable contract. CARD_BAANX_CLIENT_KEY is in the snippet; the next paragraph calls out CARD_BAANX_API_URL.

session.ts:102 — not critical (intentional)
The cache is keyed on who the token is for (host, region, email, tenant key), not on how you minted it. Password/TOTP cannot change the identity of an already-issued token. Hashing them in would force a new login whenever a string differs and fight the 429-avoidance the cache exists for. Wrong-password second calls are not this API’s job; forceRefresh / clearBaanxAuthCache exist when you want a new mint.

totp.ts:84 — not critical
After sleep we re-read the clock. To land near the end of the next 30s window, setTimeout would have to overshoot by almost a full period — not a realistic Node delay. Default period is 30s vs 2s remaining. Period 1 is a config footgun (1–300 is allowed); it is not the sandbox default. A loop would not pay for itself here.

pnpm-lock.yaml:65378 — not critical
The Baanx bits (otpauth, @noble/hashes@2.2.0) are the real adds. The Metro hunk only reshuffles transitivePeerDependencies (bufferutil / utf-8-validate) on existing metro-config snapshots — no Metro package or importer version change in this PR. Typical pnpm install annotation churn.

File: e2e/desktop/tests/utils/featureFlagUtils.ts [hunk 31]
```diff
@@ -162,6 +162,34 @@ export const FF_NEW_SEND_FLOW_FIRST_INTERACTION_BANNER_ENABLED = {
...
+ * it explicitly. `card: true` is the flag's own default param and gates the card-specific UI.
+ */
+export const FF_LWD_PAY_TAB = {
+  lwdPayTab: {
+    enabled: true,
+    params: {
+      card: true,
+      baanx_login_manifest_id: "baanx-login-url-stg",
+      baanx_hosted_manifest_id: "baanx-hosted-url-stg",
+    },
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/desktop/tests/utils/featureFlagUtils.ts:170]: fixed

File: e2e/tooling/baanx-test-client/src/auth/login.ts [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+): Promise<BaanxAuthSession> {
+  const userId = asString(asRecord(first.body).userId);
+  if (!userId) {
+    throw new BaanxOtpError(
+      "Baanx asked for an OTP but returned no userId, so the challenge cannot be triggered.",
+      redactBody(first.body, secretsOf(config)),
+    );
+  }
+
+  await triggerOtp(config, userId, fetchImpl, deps.requestTimeoutMs);
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/login.ts:76]: pushed!

File: e2e/tooling/baanx-test-client/src/http/body.ts [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+        key,
+        REDACTED_KEYS.has(key.toLowerCase()) ? REDACTION_PLACEHOLDER : redactBody(entry, secrets),
+      ]),
+    );
+  }
+
+  // Field names only get you so far: a proxy echoing a credential in `message`,
+  // `detail` or the `nonJsonBody` excerpt lands in a field we do not know, and
+  // that body is attached to errors and printed by the CLI.
+  if (typeof value === "string") return redactSecretsInText(value, secrets);
```
Comment (abdurrahman-ledger) [inline e2e/tooling/baanx-test-client/src/http/body.ts:96]: Why for strings only, can't a secret be a number for instance ?

Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/http/body.ts:96]: It isn't likely, however, I have pushed a change to account for this too.

File: e2e/tooling/baanx-test-client/src/auth/login.ts [hunk 34]
```diff
@@ -0,0 +1,225 @@
...
+
+  return {
+    accessToken,
+    userId: asString(payload.userId),
+    issuedAt: issuedAt.toISOString(),
+    expiresAt: expiresAt.toISOString(),
+    expirySource: source,
+    otpUsed,
+    verificationState: asString(payload.verificationState),
+    isLinked: asBoolean(payload.isLinked),
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/auth/login.ts]: fix pushed...

**Review by abdurrahman-ledger — DISMISSED** [review]

File: e2e/tooling/baanx-test-client/src/http/body.ts [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+
+/**
+ * Whether Baanx's own 401 message reads as a lock rather than a bad password.
+ *
+ * A guess at free text we do not control, so it only ever *adds* detail to
+ * `BaanxInvalidCredentialsError` — a miss still reports invalid credentials.
+ */
+export function looksAccountLocked(message: string | null): boolean {
+  if (!message) return false;
+  return /lock|too many (failed )?attempts|temporarily (disabled|blocked)/i.test(message);
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply e2e/tooling/baanx-test-client/src/http/body.ts]: pushed!

File: pnpm-lock.yaml [hunk 36]
```diff
@@ -65320,7 +65373,9 @@ snapshots:
     optionalDependencies:
       metro-transform-worker: 0.83.7
     transitivePeerDependencies:
+      - bufferutil
       - supports-color
+      - utf-8-validate
```
[the remark this answers is not in the document]
Comment (martijnhjk) [reply pnpm-lock.yaml:65327]: have pushed several lockfile changes but always some churn coming up - ignoring this as CI is coming back fine.

File: .changeset/pay-card-session-bootstrap.md [hunk 37]
```diff
@@ -0,0 +1,9 @@
+---
+"ledger-live-desktop": patch
```
Comment (thesan) [inline .changeset/pay-card-session-bootstrap.md]: Do you mean this patch?

**Review by abdurrahman-ledger — DISMISSED** [review]

**Review by thesan — APPROVED** [review]

File: apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts
Comment (mcayuelas-ledger) [inline apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts:1]: I think we should create a dedicated package for this, so we can share it with mobil, something like `@support/pay-card-e2e`

wdyt @thesan ?

**Review by mcayuelas-ledger — APPROVED** [review]

File: apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts
Comment (martijnhjk) [reply apps/ledger-live-desktop/src/renderer/bootstrapCardSession.ts:1]: Keeping this in mind when I look at the mobile session next 🤞

**Review by abdurrahman-ledger — APPROVED** [review]
